docs/DeferredSignals.md

```diff
@@ -0,0 +1,156 @@
+We then immediately return from the signal handler so that FEX can resume its "uninterruptible" code section without breaking anything.
+Once the "uninterruptible" code section finishes, it will store to the page, and a SIGSEGV will be raised.
+
+This is the point in which the deferred signal from before will now be handled. Inside of the SIGSEGV handler FEX will process the signal as if it was
+the previous signal that was deferred.
+- Replacing the SIGSEGV signal number with the previous captured signal number
+- Replacing the `siginfo_t` with the previous captured `siginfo_t`
+- mprotect the thread local `DeferredSignalHandlerPagePtr` to be RW.
+   - This is so that future signals get deferred, but we don't block forward code progress.
+- TODO: Overwrite mcontext_t things? I don't think this matters but there will be some private state that might leak SIGSEGV information?
+
+Once we are now handling the guest signal, FEX-Emu is in a vulnerable state where any signals received will be deferred /and/ not handled at the end
+of "uninterruptible" code sections. This is because FEX is now currently in a guest signal frame and we need to handle code compiling and other
+potentially awkward interactions without checking for additional signals.
+
+Once a guest signal handler has finished what it was working on, it will call `rt_sigreturn` or `sigreturn` which triggers FEX's SIGILL signal
+handler.
+Inside of this SIGILL signal handler FEX will restore the state of FEX /back/ to where the deferred signal handler started (The str xzr, [x0]).
+Inside of this signal handler FEX will check to see if all deferred signals are handled.
+- Checks the reference counter to see if it is zero or not.
+   - If all deferred signals are handled, mprotect the fault page back to RW
+   - If deferred signals are /not/ handled, mprotect the fault page to `PROT_NONE`
+
+Once FEX gets back to the page store, it will trampoline back to the SIGSEGV handler if it has more signals to handle.
+   - This is an edge case where we aren't expecting multiple signals in almost all cases
+   - Slightly more expensive is fine in this case.
+
+## Disadvantages of cooperative signal deferring
+Still thinking about this, come back to me. I'm concerned about signal queueing.
+- How do we handle the guest doing a long-jump out of a signal frame and still receiving signals?
+   - This will block FEX from handling /any/ more deferred signals.
+   -  Do we need to store guest stack as well to see if it has reset its own stack frame?
+   - moon-buggy does this as an example
+   - We currently just leak stack for every guest signal handler that long jumps out of the signal frame.
+      - Long term this would exhaust our stack and then crash.
+      - Test with a second guest thread where our host will only have an 8MB stack instead of the 128MB primary stack.
+      - See issue #2487
+## Expectations and considerations
+### What happens with a race condition with the refcounter?
```

Comment:
Thanks for adding this section - while reading the previous paragraphs I've been trying to convince myself this is safe, drawing the same conclusions you've made here. It's good to have the reasoning spelt out clearly!